### ISSUE TITLE
how to build graph model base edge_attr 

### ISSUE THREAD

**Issue by scalaboy** [pyg-team/pytorch_geometric#3416]
hi,all. I have a graph dataset which edge_attr is importance. where I can't get a example about how to train a graph base edge_attr .like mnist_nn_conv.py is not solver my problem

**Comment by rusty1s** [comment]
In general, you can utilize any GNN layer that can handle `edge_attr` as input, see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/cheatsheet.html), *e.g.*, the [`GINEConv`](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.GINEConv). You can simply call your layer via
```python
x = conv(x, edge_index, edge_attr)
```
Does that resolve your concerns?

**Comment by for-just-we** [comment]
> In general, you can utilize any GNN layer that can handle `edge_attr` as input, see [here](https://pytorch-geometric.readthedocs.io/en/latest/notes/cheatsheet.html), _e.g._, the [`GINEConv`](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.conv.GINEConv). You can simply call your layer via
> 
> ```python
> x = conv(x, edge_index, edge_attr)
> ```
> 
> Does that resolve your concerns?

hello, would other conv layers API like `GatedGraphConv`, `GCNConv`  support this `edge_attr` parameter? 
It seems that in DGL library, `GatedGraphConv` support `edge_attr`.

**Comment by rusty1s** [comment]
I just looked into the DGL version of `GatedGraphConv` and it does not look like they support edge features either.

Nonetheless, I'm more than happy to let more GNN layers support edge features. How would you like to add edge feature support to `GatedGraphConv`?

**Comment by for-just-we** [comment]
> I just looked into the DGL version of `GatedGraphConv` and it does not look like they support edge features either.
> 
> Nonetheless, I'm more than happy to let more GNN layers support edge features. How would you like to add edge feature support to `GatedGraphConv`?

I learn this from [devign model](https://github.com/saikat107/Devign/blob/master/modules/model.py), the code is for source code vulnerability detection task by graph network, which is also what I'm doing.

In the vulnerability detection task. There are different types of edge in a graph. For example, control flow edge, data dependence edge. i.e, I hope to use `edge_attr` to denote which type of edge it is. Maybe `edge_attr` is a simple one-hot vector. If there are 4 types of edge. CFG edge, CPG edge, DDG edge, AST edge. [1, 0, 0, 0] represents the edge is CFG edge and [0, 1, 0, 0] represents CPG edge and so on.

In the code, the author define a `GatedGraphConv` by `self.ggnn = GatedGraphConv(in_feats=input_dim, out_feats=output_dim,
n_steps=num_steps, n_etypes=max_edge_types)` and use like `outputs = self.ggnn(graph, features, edge_types)`.  There is an additional parameter `edge_types` passed into ggnn layer.

I haven't used DGL yet. For now, I would like to support different types of edge by `edge_attr` because I haven't seen another way.

**Comment by rusty1s** [comment]
I see. Thanks for clarifying. You can either integrate it that by simply concatenating the edge features to `x_j`, e.g.:
```python
def message(self, x_j, edge_attr):
    return self.lin(torch.cat([x_j, edge_attr], dim=-1))
```
or by using edge types (similar to how DGL is doing it), e.g., via our [`HeteroLinear`]() layer:
```python
self.hetero_lin = HeteroLinear(in_channels, out_channels, num_types=num_edge_types)

def message(self, x_j, edge_type):
    return self.hetero_lin(x_j)
```

**Comment by rusty1s** [comment]
Yes, exactly :)

**Comment by for-just-we** [comment]
> Yes, exactly :)

OK, thank you very much

**Comment by scalaboy** [comment]
> 

I need an example,contains dataset,run fit code,train and test etc.
not a api.....

**Comment by rusty1s** [comment]
Can you clarify? What about [this one](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/pna.py)?

**Comment by suice07** [comment]
> I see. Thanks for clarifying. You can either integrate it that by simply concatenating the edge features to `x_j`, e.g.:
> 
> ```python
> def message(self, x_j, edge_attr):
>     return self.lin(torch.cat([x_j, edge_attr], dim=-1))
> ```
> 
> or by using edge types (similar to how DGL is doing it), e.g., via our [`HeteroLinear`](https://pytorch-geometric.readthedocs.io/en/latest/modules/nn.html#torch_geometric.nn.dense.HeteroLinear) layer:
> 
> ```python
> self.hetero_lin = HeteroLinear(in_channels, out_channels, num_edge_types)
> 
> def message(self, x_j, edge_type):
>     return self.hetero_lin(x_j, edge_type)
> ```

Hello dear rusty1s,
   I 'd like to implement my algorithm also by concate the aggregated node feature with the edge feature. Like you mentioned before,  ```torch.cat([x_j, edge_attr]``` can be used, but how should I store the edge_attr? and I am strill confusing that x_j should be node feature of node j right?

**Comment by rusty1s** [comment]
It should be `return torch.cat([x_j, edge_attr], dim=-1)`. Since `edge_attr` is already an edge-level tensor, there is no need to appeng `_i` to it. Now both `x_j` and `edge_attr` will have shape `[num_edges, *]`.

**Comment by suice07** [comment]
> torch.cat([x_j, edge_attr], dim=-1)

I have the x in shape [25,15],25 is the node number, edge_attr in shape[56,6],56 is the number of edges, and using out = self.propagate(edge_index, x=node, norm=norm, edge_attr = edge_attr) to pass the message, in the message function, the shape of x_j is [81,15],but edge_attr is still[56,6],seems that it passed the whole edge_attr to message function instead of those of a specific edge. did I miss some steps?

**Comment by rusty1s** [comment]
It looks like your GNN layer adds self-loops to the graph, that's why you have `x_j.size(0) = 25 + 56 = 81`. Can you remove this logic?

**Comment by suice07** [comment]
> It looks like your GNN layer adds self-loops to the graph, that's why you have `x_j.size(0) = 25 + 56 = 81`. Can you remove this logic?

oh,that's the point, so x_j is the aggregated message from node j, then concatenate it with the edge_attr through ```torch.cat([x_j, edge_attr], dim=-1)```
Thanks so much! I think it is solved

**Comment by rusty1s** [comment]
It is only 56 in case you have 56 edges. The first dimension can be seen as the batch dimension here, as we share model weights across all edges.

**Comment by suice07** [comment]
> It is only 56 in case you have 56 edges. The first dimension can be seen as the batch dimension here, as we share model weights across all edges.
Thanks for reply, dear @rusty1s ,
nope，I mean that I have a parallel network,the first one takes node feature as x,which propagate like this:```self.propagate(edge_index, x=node_feature, edge_attr = edge_feature)```,the other one switches node feature and edge feature,in my case takes nodes as edges,edges as nodes,which propagates like ``` self.propagate(node_index, x=edge_feature, edge_attr = node_feature)```,node feature has a dimension [25,m],edge feature has a dimension of [28,n],but when I print the ```x_j``` in the message function, in both network, the dimension of ```x_j``` is [56,t]

**Comment by rusty1s** [comment]
For a edge network, you can take a look at `transforms.LineGraph` and compare it to your implementation. It will treat edges as nodes and connects two nodes whenever there exists a node that connects the two edges in the original network. Let me know!

**Comment by suice07** [comment]
> For an edge network, you can take a look at `transforms.LineGraph` and compare it to your implementation. It will treat edges as nodes and connects two nodes whenever there exists a node that connects the two edges in the original network. Let me know!

the original data is like ```Data(edge_index=[2, 56], edge_attr=[56, 6], node_feature=[25, 27], smiles='N[C@@H](Cc1c[nH]c2ccccc12)C(=O)Nc1ccc2ccccc2c1')```
I called LineGraph ``` data_edge = transform(data)```,then the new data is turned to:
```Data(edge_index=[2, 134], node_feature=[25, 27], smiles='N[C@@H](Cc1c[nH]c2ccccc12)C(=O)Nc1ccc2ccccc2c1', x=[56, 6], num_nodes=56)```
According to the adjacency matrix I caculated before ,I think the new edge_index is right. but I am confused to perform my algorithm throught this.

what I want to do is for a specific edge from node a to node b, the edge ab takes all aggregated edge features from the edges which connected to start node a, attached by its start node feature; 

but for the transformed data,it takes edge as node,but  according to the edge_index of it, the transformed 'edge' seems to have nothing to do with the real node

**Comment by rusty1s** [comment]
Isn't that what is already happening? If you also want to integrate node features into propagation, you can convert them to edge features first:
```python
row, col = data.edge_index
data.edge_attr = torch.cat([data.x[row], data.x[col], data.edge_attr], dim=-1)
data = transform(data)
```

**Comment by suice07** [comment]
> Isn't that what is already happening? If you also want to integrate node features into propagation, you can convert them to edge features first:
> 
> ```python
> row, col = data.edge_index
> data.edge_attr = torch.cat([data.x[row], data.x[col], data.edge_attr], dim=-1)
> data = transform(data)
> ```

I think I got the point, thanks for all the help!
